### PR TITLE
Implement MPI

### PR DESCRIPTION
**Pull request by vchuravy**

### CODE REVIEW

File: src/comm.jl [hunk 1]
```diff
@@ -1,264 +1,1887 @@
...
+   #   domain->recvRequest[i] = MPI_REQUEST_NULL ;
+   # }
+
+   myRank = MPI.Comm_rank(domain.comm)
 
    # post receives
+   function irecv!(fromProc, offset, recvCount)
+      idx = offset + 1
+      data = view(domain.commDataRecv, idx:(idx+recvCount))
+      return MPI.Irecv!(data, fromProc, msgType + fromProc, MPI.MPI_COMM_WORLD)
```
Comment (wsmoses) [inline src/comm.jl:46]: Adding fromproc is different: https://github.com/wsmoses/Enzyme-MPI/blob/d211d3522628628cc3990a2470639b3159104a37/lulesh-comm.cc#L112

File: src/comm.jl [hunk 2]
```diff
@@ -1,264 +1,1887 @@
...
          emsg += 1
       end
 
-      if (rowMax && colMin)
-          fromRank = myRank + domain.m_tp - 1
-        start = pmsg * maxPlaneComm + emsg * maxEdgeComm
-      buf = @view domain.commDataRecv[start:start + dz *xferFields]
-      domain.recvRequest[pmsg+emsg] = MPI.Irecv!(buf, fromRank, msgType, comm)
+      if rowMax && colMin
+         fromProc = myRank + domain.tp + 1
```
Comment (wsmoses) [inline src/comm.jl:161]: Minus one for fromProc: https://github.com/wsmoses/Enzyme-MPI/blob/d211d3522628628cc3990a2470639b3159104a37/lulesh-comm.cc#L219

File: src/comm.jl [hunk 3]
```diff
@@ -1,264 +1,1531 @@
...
+   packable = true
+   # TODO: What is this checking for?
+   # Probably that fieldData is all the same
+   for i = 1:(xferFields - 2)
+       if fieldData[i+1] - fieldData[i] !=
+          fieldData[i+2] - fieldData[i+1]
+           packable = false
+           break
+       end
+   end
```
Comment (wsmoses) [inline src/comm.jl:328]: ??

File: src/comm.jl [hunk 4]
```diff
@@ -1,264 +1,1531 @@
...
+               error("unimplemented")
+            end
+
+            otherRank = myRank + domain.tp
+            req = MPI.Isend(src, otherRank, msgType + myRank, domain.comm)
+            domain.sendRequest[pmsg+1] = req
+            pmsg += 1
+         end
+
+         if rowMax && doSend
```
Comment (wsmoses) [inline src/comm.jl:478]: This is a repeat of the previous case of rowMax && doSend.

File: src/comm.jl [hunk 5]
```diff
@@ -1,264 +1,1531 @@
...
+            else
+               error("unimplemented")
+            end
+
+            otherRank = myRank + domain.tp
+            req = MPI.Isend(src, otherRank, msgType + myRank, domain.comm)
+            domain.sendRequest[pmsg+1] = req
+            pmsg += 1
+         end
+
```
Comment (wsmoses) [inline src/comm.jl:500]: This colMin | colMax should not be inside of the row case.

File: src/comm.jl [hunk 6]
```diff
@@ -1,264 +1,1531 @@
...
+            if colMin
+               # contiguous memory
+               if packCol
+                  srcOffset = 0
+                  offset = pmsg * maxPlaneComm
+                  for field in fields
+                     for i in 0:(dz-1)
+                        for j in 0:(dy-1)
+                           # INDEXING +1 missign somewhere
+                           domain.commDataSend[offset+i*dx+j] = field[srcOffset+i*dx*dy+j*dx]
```
Comment (wsmoses) [inline src/comm.jl:544]: Field requires a +1 for 1-indexed, dest addr is i*DY not i*dx, https://github.com/wsmoses/Enzyme-MPI/blob/d211d3522628628cc3990a2470639b3159104a37/lulesh-comm.cc#L503

File: src/comm.jl [hunk 7]
```diff
@@ -1,264 +1,1531 @@
...
+            if colMax && doSend
+               # contiguous memory
+               if packCol
+                  srcOffset = dx - 1
+                  offset = pmsg * maxPlaneComm
+                  for field in fields
+                     for i in 0:(dz-1)
+                        for j in 0:(dy-1)
+                           # INDEXING +1 missign somewhere
+                           domain.commDataSend[offset+i*dy+j] = field[srcOffset+j*dx]
```
Comment (wsmoses) [inline src/comm.jl:579]: Field needs +1 for 1 index, also missing the i*dx*dy term: https://github.com/wsmoses/Enzyme-MPI/blob/d211d3522628628cc3990a2470639b3159104a37/lulesh-comm.cc#L522

File: src/comm.jl [hunk 8]
```diff
@@ -1,264 +1,1531 @@
...
+                  src = view(domain.commDataSend, idx:(idx+(xferFields * sendCount)))
+               else
+                  error("unimplemented")
+               end
+               otherRank = myRank + 1
+               req = MPI.Isend(src, otherRank, msgType + myRank, domain.comm)
+               domain.sendRequest[pmsg+1] = req
+               pmsg += 1
+            end
+
```
Comment (wsmoses) [inline src/comm.jl:594]: This plane only down should be a top level if.

File: src/comm.jl [hunk 9]
```diff
@@ -1,264 +1,1531 @@
...
+         if !planeOnly
+            if rowMin && colMin
+               offset = pmsg * maxPlaneComm + emsg * maxEdgeComm
+               for field in fields
+                  for i in 0:(dz-1)
+                     domain.commDataSend[offset+i] = field[i*dx*dy]
```
Comment (wsmoses) [inline src/comm.jl:600]: CommDataSend and field require +1 for 1-indexing,

File: src/comm.jl [hunk 10]
```diff
@@ -1,264 +1,1531 @@
...
+               req = MPI.Isend(src, otherRank, msgType + myRank, domain.comm)
+               domain.sendRequest[pmsg+emsg+1] = req
+               emsg += 1
+            end
+
+            if rowMin && planeMin
+               offset = pmsg * maxPlaneComm + emsg * maxEdgeComm
+               for field in fields
+                  for i in 0:(dx-1)
+                     domain.commDataSend[offset+i] = field[i]
```
Comment (wsmoses) [inline src/comm.jl:617]: +1 to both commDataSend and field index.

File: src/comm.jl [hunk 11]
```diff
@@ -1,264 +1,1531 @@
...
+            domain.sendRequest[pmsg+emsg+1] = req
+            emsg += 1
+         end
+
+         if rowMax && planeMax && doSend
+            offset = pmsg * maxPlaneComm + emsg * maxEdgeComm
+            srcOffset = dx*(dy-1) + dx*dy*(dz-1)
+            for field in fields
+               for i in 0:(dx-1)
+                  domain.commDataSend[offset+i] = field[srcOffset+i]
```
Comment (wsmoses) [inline src/comm.jl:667]: +1 to both indices,

File: src/comm.jl [hunk 12]
```diff
@@ -1,264 +1,1531 @@
...
+            domain.sendRequest[pmsg+emsg+1] = req
+            emsg += 1
+         end
+
+         if colMax && planeMax && doSend
+            offset = pmsg * maxPlaneComm + emsg * maxEdgeComm
+            srcOffset = dx*dy*(dz-1) + dx
+            for field in fields
+               for i in 0:(dy-1)
+                  domain.commDataSend[offset+i] = field[srcOffset+i*dx]
```
Comment (wsmoses) [inline src/comm.jl:684]: commDataSend (not field) require +1, I would however move the -1 (implicitly deleted) back into srcOffset and add an explicit +1 for consistency sake.

File: src/comm.jl [hunk 13]
```diff
@@ -1,264 +1,1531 @@
...
+            domain.sendRequest[pmsg+emsg+1] = req
+            emsg += 1
+         end
+
+         if rowMax && colMin && doSend
+            offset = pmsg * maxPlaneComm + emsg * maxEdgeComm
+            srcOffset = dx*(dy-1)
+            for field in fields
+               for i in 0:(dz-1)
+                  domain.commDataSend[offset+i] = field[srcOffset+i*dx*dy]
```
Comment (wsmoses) [inline src/comm.jl:701]: Both sides require +1

File: src/comm.jl [hunk 14]
```diff
@@ -1,264 +1,1531 @@
...
+            domain.sendRequest[pmsg+emsg+1] = req
+            emsg += 1
+         end
+
+         if rowMin && colMax
+            offset = pmsg * maxPlaneComm + emsg * maxEdgeComm
+            srcOffset = dx - 1
+            for field in fields
+               for i in 0:(dz-1)
+                  domain.commDataSend[offset+i] = field[srcOffset+i*dx*dy]
```
Comment (wsmoses) [inline src/comm.jl:752]: Both sides require +1 to index.

File: src/comm.jl [hunk 15]
```diff
@@ -1,264 +1,1531 @@
...
+            domain.sendRequest[pmsg+emsg+1] = req
+            emsg += 1
+         end
+
+         if rowMax && planeMin
+            offset = pmsg * maxPlaneComm + emsg * maxEdgeComm
+            srcOffset = dx*(dy - 1)
+            for field in fields
+               for i in 0:(dx-1)
+                  domain.commDataSend[offset+i] = field[srcOffset+i]
```
Comment (wsmoses) [inline src/comm.jl:769]: Both sides require +1 to index.

File: src/comm.jl [hunk 16]
```diff
@@ -1,264 +1,1531 @@
...
+            domain.sendRequest[pmsg+emsg+1] = req
+            emsg += 1
+         end
+
+         if colMax && planeMin
+            offset = pmsg * maxPlaneComm + emsg * maxEdgeComm
+            srcOffset = dx - 1
+            for field in fields
+               for i in 0:(dy-1)
+                  domain.commDataSend[offset+i] = field[srcOffset+i*dx]
```
Comment (wsmoses) [inline src/comm.jl:786]: Both sides require +1 to index.

File: src/comm.jl [hunk 17]
```diff
@@ -1,264 +1,1531 @@
...
+            end
+            idx = pmsg * maxPlaneComm + emsg * maxEdgeComm + cmsg * CACHE_COHERENCE_PAD_REAL + 1
+            src = view(domain.commDataSend, idx:(idx+(xferFields * 1)))
+            otherRank = myRank - domain.tp^2 - domain.tp - 1
+            req = MPI.Isend(src, otherRank, msgType + myRank, domain.comm)
+            domain.sendRequest[pmsg+emsg+cmsg+1] = req
+            cmsg += 1
+         end
+
+
```
Comment (wsmoses) [inline src/comm.jl:812]: Obviously missing from https://github.com/wsmoses/Enzyme-MPI/blob/d211d3522628628cc3990a2470639b3159104a37/lulesh-comm.cc#L767 to end

File: src/comm.jl (hunk 1 above)
Comment (vchuravy) [reply src/comm.jl:46]: Seeing them side by side I withdraw my objection. The transfer is already tagged from which rank it is coming and thus does not need `+fromProc` mixing.

File: src/comm.jl [hunk 18]
```diff
@@ -1,264 +1,1531 @@
...
+            pmsg += 1
+         end
+         if rowMax && doSend
+            # contiguous memory
+            if packRow
+               srcOffset = dx*(dy - 1)
+               offset = pmsg * maxPlaneComm
+               for field in fields
+                  for i in 0:(dz-1)
+                     copyto!(domain.commDataSend, offset+i*dx , field, srcOffset+i*dx*dy, dx)
```
[the remark this answers is not in the document]
Comment (vchuravy) [reply src/comm.jl:461]: That's sourceOffset itself see a few lines above

[the remark this answers is not in the document]
Comment (wsmoses) [reply src/comm.jl:461]: You're right it's there now, not sure what happened.

File: src/comm.jl (hunk 8 above)
Comment (vchuravy) [reply src/comm.jl:461]: I think I corrected that now.

File: src/comm.jl (hunk 12 above)
Comment (vchuravy) [reply src/comm.jl]: Not sure I follow, why does field not require `+1`?

File: src/comm.jl (hunk 17 above)
Comment (vchuravy) [reply src/comm.jl:741]: I seemingly forgot to push

File: src/comm.jl [hunk 19]
```diff
@@ -1,264 +1,1362 @@
...
+         idx = pmsg * maxPlaneComm + emsg * maxEdgeComm + cmsg * CACHE_COHERENCE_PAD_REAL + 1
+         src = view(domain.commDataSend, idx:(idx+xferFields))
+         otherRank = myRank + domain.tp^2 + domain.tp + 1
+         req = MPI.Isend(src, otherRank, msgType, domain.comm)
+         domain.sendRequest[pmsg+emsg+cmsg+1] = req
+         cmsg += 1
+      end
+   end
+
+   resize!(domain.sendRequest, pmsg+emsg+cmsg)
```
Comment (wsmoses) [inline src/comm.jl:774]: Can we do a waitall that takes a view of appropriate size rather than resizing?

File: src/comm.jl [hunk 20]
```diff
@@ -1,264 +1,1579 @@
...
+      opCount = dy * dz
+
+      if colMin && doRecv
+         # contiguous memory
+         MPI.Wait!(domain.recvRequest[pmsg+1])
+         offset = pmsg * maxPlaneComm
+         for field in fields
+            for i in 0:(dz-1)
+               for j in 0:(dy-1)
+                  field[i*dx*dy+j*dx] = domain.commDataRecv[offset + i*dy + j]
```
Comment (wsmoses) [inline src/comm.jl:1314]: Off by ones

**Pull request merged**